teps/draft-20200911-hermekton.md

```diff
@@ -0,0 +1,214 @@
+
+A company wishes to apply policy on what software is used as a dependency during their build process.
+They can then define all dependencies as inputs or steps, and run builds in a hermetic environment to ensure nothing else is pulled in by accident.
+
+Note: This is very similar to Google's internal requirements for build processes.
+
+### Notes/Constraints/Caveats (optional)
+
+The isolation boundary offered by this technique is based on Linux namespaces, the same technologies that are used for Linux containers.
+That means that these are not a full security boundary.
+Hermetic builds should not be used to run otherwise untrusted code.
+The privileges and capabilities used in your TaskRun containers will also be present in your hermetic builds.
+Container escapes (and then network access) will always be possible.
+
+Hermetic builds should only be used to help detect and prevent accidental network access, and as an extra layer of defense against insider attacks. 
+These need not stop a determined adversary.
+
+### Risks and Mitigations
+
+One potential risk is that this namespace-level sandboxing is insufficient to make guarantees around hermeticity for most potential users of this feature.
+We can mitigate this by surveying users ahead of time, and we can always add more "strict" execution modes down the road.
+Note: We have verified that this will meet Google's requirements for hermetic builds of otherwise trusted code.
+
+
+### User Experience (optional)
+
+See the API changes outlined above.
+
+
+### Performance (optional)
+
+No performance implications are expected.
+
+## Design Details
+
+### Implementation
+
+See Alternatives Considered for a discussion of alternative approaches.
+
+The Tekton Pipelines entrypoint binary will be extended to support executing hermetic builds.
+This will consist of executing the user-controlled container entrypoint in a new Linux network namespace that is not configured.
+
+This requires passing some flags to `cmd.SysProcAttr.CloneFlags` before calling cmd.Run. 
+We will pass:
+
+```
+syscall.CLONE_NEWNET | syscall.CLONE_NEWUSER | syscall.CLONE_NEWNS | syscall.CLONE_NEW_PID
+```
+
+`CLONE_NEWNET` is the main flag we care about.
+This gives us a new, empty network namespace that will not work.
+To use this flag inside a Kubernetes container, we must also create a new user namespace using the other flags (unless the container runs with `privileged: True`). 
+
+After creating the user namespace, we must also map in the external users and groups.
+That can be done with the `cmd.SysProcAttr.UidMappings` and `cmd.SysProcAttr.GidMappings` fields, respectively. 
+
+The net result is a process running in a namespace with an identical filesystem, the same users and groups, and no networking.
+
+The entrypointer will need to know when to drop networking - we don't want to do this on every container!
+Specifically, init and resource container steps should still run with networking before and after the Task steps.
```

Comment:
_Should_ init and system containers run with networking? Off the top of my head I don't think any of them need it today, except for source-fetching steps prepended by `PipelineResource`s -- if we assume/require users don't use resources and instead fetch in a previous Task and pass in through `workspaces`, I think all of this becomes much simpler and harder to get wrong.

Reply:
Debateable, but I think yes. I'm paraphrasing from our internal requirements to make them make sense outside (there's nothing secret, it's just full of jargon):

> The build process is either disallowed to fetch source dependencies, or must fetch source
dependencies in a manner that’s traceable to the builder.

By having the build system (the init containers in today's world) do the fetching, we get strong, trustable provenance. Having other Tasks in the same Pipeline fetch dependencies just moves the trust problem to another location. The entire Pipeline needs to be hermetic.

This could all be solved by building a higher level build system that uses another DSL that compiles down to "System Tasks" and "User Tasks", and runs the "User Tasks" without networking, but I'm trying to avoid that.

Reply:
It also doesn't really change the mechanism - it would just remove an `if` statement in the entrypoint code I think. NetworkPolicies work at the Pod-level, but they're way too finicky to rely on IMO. After spending a day debugging them on GKE I'd rather not use them at all.